File: apps/web/src/lib/api/gateway.ts
/**
 * Gateway API client — OpenClaw gateway health, status, and messaging.
 */
import { apiFetch } from "./client";

// ── Types ─────────────────────────────────────────────────

export interface GatewayHealth {
  connected: boolean;
  status: "healthy" | "unhealthy" | "unreachable" | "timeout" | "error" | "not_configured";
  latency_ms?: number;
  gateway_url?: string;
  version?: string;
  uptime?: string;
  agents_loaded?: number;
  http_status?: number;
  error?: string;
  checked_at?: string;
  mock_mode?: boolean;
}

export interface GatewayAgent {
  id: string;
  name: string;
  status: string;
  model: string | null;
  workspace?: string;
  channels: string[];
  is_default: boolean;
}

export interface GatewaySession {
  id: string;
  agent_id: string | null;
  status: string;
  created_at: string | null;
  last_activity: string | null;
  message_count: number | null;
}

export interface ChecklistItem {
  id: string;
  label: string;
  status: "pass" | "fail" | "warn" | "skip";
  detail: string;
}

export interface GatewayStatus {
  health: GatewayHealth;
  agents: GatewayAgent[];
  sessions: GatewaySession[];
  checklist: ChecklistItem[];
  mock_mode?: boolean;
  config: {
    gateway_url_set: boolean;
    gateway_token_set: boolean;
    agent_api_key_set: boolean;
    openai_key_set: boolean;
  };
}

export interface GatewayMessageResponse {
  session_id?: string;
  response?: string;
  status?: string;
  [key: string]: unknown;
}

// ── API Methods ───────────────────────────────────────────

export const gatewayApi = {
  /** Check gateway connectivity. */
  health: () => apiFetch<GatewayHealth>("/gateway/health"),

  /** Full deployment status (health + agents + sessions + checklist). */
  status: () => apiFetch<GatewayStatus>("/gateway/status"),

  /** List agents from gateway. */
  agents: () => apiFetch<GatewayAgent[]>("/gateway/agents"),

  /** List active sessions on gateway. */
  sessions: () => apiFetch<GatewaySession[]>("/gateway/sessions"),

  /** Send a message to an agent via gateway. */
  sendMessage: (agentId: string, message: string, sessionId?: string) =>
    apiFetch<GatewayMessageResponse>("/gateway/message", {
      method: "POST",
      body: JSON.stringify({
        agent_id: agentId,
        message,
        session_id: sessionId,
      }),
    }),
};

File: apps/api/app/routers/gateway.py
"""Gateway API: proxy endpoints for OpenClaw gateway communication.

Endpoints:
  GET  /gateway/health   — Check gateway connectivity
  GET  /gateway/status   — Full deployment status (health + agents + checklist)
  GET  /gateway/agents   — List agents from gateway
  POST /gateway/message  — Send message to an agent via gateway
  GET  /gateway/sessions — List active gateway sessions
"""

from __future__ import annotations

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.auth import get_current_user, CurrentUser
from app.services.gateway_client import (
    check_health,
    get_full_status,
    list_gateway_agents,
    get_gateway_sessions,
    send_message_to_agent,
    GatewayError,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/gateway", tags=["gateway"])


# ── Request/Response Models ──────────────────────────────────

class GatewayMessageRequest(BaseModel):
    agent_id: str = Field(..., min_length=1, max_length=80, pattern=r"^[a-zA-Z0-9_-]+$")
    message: str = Field(..., min_length=1, max_length=10000)
    session_id: Optional[str] = Field(None, max_length=200)


# ── Endpoints ────────────────────────────────────────────────

@router.get("/health")
async def gateway_health(user: CurrentUser = Depends(get_current_user)):
    """Check if the OpenClaw gateway is reachable and healthy."""
    try:
        return await check_health()
    except GatewayError as e:
        return {
            "connected": False,
            "status": "not_configured",
            "error": str(e),
        }


@router.get("/status")
async def gateway_status(user: CurrentUser = Depends(get_current_user)):
    """Full deployment status: health + agents + sessions + checklist.

    This is the main endpoint for the Deployment Dashboard.
    """
    try:
        return await get_full_status()
    except GatewayError as e:
        return {
            "health": {
                "connected": False,
                "status": "not_configured",
                "error": str(e),
            },
            "agents": [],
            "sessions": [],
            "checklist": [],
            "config": {
                "gateway_url_set": False,
                "gateway_token_set": False,
                "agent_api_key_set": False,
                "openai_key_set": False,
            },
        }


@router.get("/agents")
async def gateway_agents(user: CurrentUser = Depends(get_current_user)):
    """List agents known to the OpenClaw gateway."""
    try:
        return await list_gateway_agents()
    except GatewayError as e:
        raise HTTPException(503, str(e))


@router.get("/sessions")
async def gateway_sessions(user: CurrentUser = Depends(get_current_user)):
    """List active sessions on the OpenClaw gateway."""
    try:
        return await get_gateway_sessions()
    except GatewayError as e:
        raise HTTPException(503, str(e))


@router.post("/message")
async def gateway_send_message(
    body: GatewayMessageRequest,
    user: CurrentUser = Depends(get_current_user),
):
    """Send a message to an agent via the OpenClaw gateway.

    Returns the gateway's response (session ID, agent reply, etc.).
    """
    try:
        result = await send_message_to_agent(
            agent_id=body.agent_id,
            message=body.message,
            session_id=body.session_id,
        )
        return result
    except GatewayError as e:
        raise HTTPException(
            status_code=e.status_code if e.status_code >= 400 else 503,
            detail=str(e),
        )
